"""Unit tests for vulnerability protobuf message with library name metadata."""

from src.ostorlab.agent.message.proto.v3.report.vulnerability import vulnerability_pb2
from src.ostorlab.agent.message.proto.v3.asset.file.android.apk import apk_pb2


def testMessage_whenCreateWithVulnLocationAndMetadataAsLibraryName_shouldSerializeAndDeserializeCorrectly():
    """Test that vulnerability message with metadate library name serializes correctly."""
    apk_message = apk_pb2.Message()
    apk_message.content_url = "https://api.example.com/app.apk"
    apk_message.android_metadata.package_name = "com.example.myapp"

    vulnerability_location_message = vulnerability_pb2.VulnerabilityLocation()
    vulnerability_location_message.android_apk.CopyFrom(apk_message)
    metadata = vulnerability_location_message.metadata.add()
    metadata.type = vulnerability_pb2.MetadataTypeEnum.LIBRARY_NAME
    metadata.value = "openssl"

    vulnerability_message = vulnerability_pb2.Message()
    vulnerability_message.title = "vuln title"
    vulnerability_message.vulnerability_location.CopyFrom(
        vulnerability_location_message
    )

    serialized = vulnerability_message.SerializeToString()
    deserialized = vulnerability_pb2.Message()
    deserialized.ParseFromString(serialized)

    assert deserialized.title == "vuln title"
    assert len(deserialized.vulnerability_location.metadata) == 1
    assert deserialized.vulnerability_location.metadata[0].type == 12
    assert deserialized.vulnerability_location.metadata[0].value == "openssl"
    assert (
        deserialized.vulnerability_location.android_apk.content_url
        == "https://api.example.com/app.apk"
    )
    assert (
        deserialized.vulnerability_location.android_apk.android_metadata.package_name
        == "com.example.myapp"
    )
